Refactor TodoList using redux

gringrape opened the pull request:
# To-do 리스트 Redux를 사용하여 리팩터링 하기

지금은 모든 상태를 `App` 컴포넌트가 관리하고 있습니다. `Redux`가 관리하도록 리팩터링을 해주세요.

## 과제 목표

상태관리를 `Redux`로 변경해도 모든 컴포넌트 테스트가 통과해야합니다. 추가로 `reducer`들에 테스트를 추가해주세요.

## 제한 조건

- 모든 기능은 TDD로 구현합니다. 컴포넌트와 reducer에 대한 테스트가 존재해야 합니다. 
- `if`는 사용할 수 있어도 `else` 사용하실 수 없습니다. `GuardClauses` 방법을 사용해주세요.
- `switch`는 사용하실 수 없습니다.
- `let`은 사용하실 수 없습니다. `const`만을 사용하여 문제를 해결해주세요.
- 함수 이름과 변수 이름에 줄임말은 사용하실 수 없습니다. 길더라도 명확하게 써주세요.
- indent(인덴트, 들여쓰기) depth를 1로 유지해주세요.
예를 들어, for문 안에 if문이 있으면 indent depth는 2입니다.
depth를 줄이는 좋은 방법은 함수(또는 메소드)를 분리하면 됩니다.

gringrape committed:
Add updateTaskTitle test to reducer

gringrape committed:
Implement updateTaskTitle

gringrape committed:
Refactor updateTaskTitle test, implementation

add actions.js to manage every kind of action in one place.
fix payload to object considering general case with many payloads

gringrape committed:
Add invalid title case to updateTaskTitle test

gringrape committed:
Implement invalid title case of updateTaskTitle test

gringrape commented:
## 작업내용
아직 내용파악을 완전히 못해서, 목적을 정확히 알고있고 독립적인 reducer 부터 진행해보았습니다. 
테스트를 추가하고 천천히 내용을 파악하며 진행해보겠습니다.

wholemann commented:
네 좋은 시도 같아요. UI와 상관 없는 비교적 단순하고 요구사항이 명확한 함수부터 TDD를 시도해보세요.

gringrape committed:
Fix test name for clarity

gringrape committed:
Add test to reducer

to check undefined state case

gringrape committed:
Implement reducer

to deal with undefined state case.

gringrape committed:
Add test to reducer - addTask action

gringrape committed:
Implement reducer and fix test

Implment addTask action case
Fix matcher to check object equality

gringrape committed:
Fix reducer test

in addTaskAction to test if newtitle is empty string

gringrape committed:
Fix reducer test

to check if newId is updated correctly

gringrape committed:
Refactor reducer test

Write object in one line for readability
Use variable instead of hard-coded value
Fix test name
Extract constant

gringrape committed:
Add test to reducer

to test emptyTaskTitle case

gringrape committed:
Implement reducer

to deal with blank task title case

gringrape committed:
Refactor reducer test, implementation

gringrape committed:
Add deleteTask test to reducer

gringrape committed:
Implement deleteTask case on reducer

gringrape committed:
Refactor reducer and test

gringrape committed:
Add integration test to App Component

to ensure application work correctly

gringrape committed:
Refactor Add Component to use reducer

gringrape committed:
Add test to App Component

to use redux

gringrape committed:
Install react-redux

gringrape committed:
Install redux

gringrape committed:
Refactor App Comonent

to use useSelector, useDispatch

gringrape commented:
## 작업 내용
- reducer 의 구현, 테스트 추가
- App 테스트 추가 (통합테스트)
- App 에서 store, dispatcher, reducer, action 을 이용하도록 리팩토링

## 고민지점
### 1. validation 위치
reducer 를 테스트하면서, 유효하지 않은 값이 들어오는 경우를 테스트하고 구현해보았습니다.
reducer 자체가 값검증로직도 함께 가지면서 복잡해져서, 검증로직을 함수단위마다 모두 갖는게 좋은건지
어플리케이션 흐름중에 한군데에서 모든 validation 을 수행하는게 좋을지 고민중입니다.
후자가 효율적일것 같은데, 이렇게 되면 어디서 진행해야할지 잘 모르겠습니다.

## 고칠점
### 1. reducer 테스트의 장황함.
기능에 비해 테스트가 장황한것 같습니다. 다시 보니 indent 도 많이 되어있어서
강의를 다시 보면서 좀 더 단순하게 수정하고 싶습니다.

gringrape committed:
Delete ? operator

and added default action type.
Because to use ?, I needed to install another babel-loader
I just deleted it.

gringrape committed:
Refactor reducer not to use if syntax

gringrape committed:
Fix reducer to remove unnecessary function call

gringrape committed:
Refactor reducer to extract only what needed to be extracted

update has the same meaning with reducer
so it doesn't need to extract it out

gringrape committed:
Add testcases to reducer test

gringrape commented:
## 작업내용
1. reducer 를 if 문을 사용하지 않도록 리팩토링
## 작업후기
### 변경의 편리함
상세하게 테스트 코드가 나눠져 있어서 디버깅을 따로 하지 않아도 어떤 문제가 발생되었는지
발견하기 편했습니다. 자신있게 변경을 할 수 있어서 좋았습니다.
### 함수 추출시 불편함
하나의 함수를 여러함수로 쪼개다 보니,
함수가 어떤 타입을 받아서 어떤 타입을 내놓는지 호출할때도
간편하게 확인할 수 있었으면 더 좋았겠다는 생각이 듭니다.

wholemann commented:
hooking이 아니라 mocking입니다.

wholemann commented:
말씀하신 부분 때문에 TypeScript를 사용하죠. 특히 리덕스 쪽은 타입을 잡아주면 좋은 것은 맞습니다. UI 쪽에는 굳이 써봤자 고통이라고 봅니다만ㅎㅎ

wholemann commented:
함수 validation 같은 경우 일단은 각 함수에서 합니다. 다만 공통 validation이 보일 경우 그때 분리하면서 정책을 생각해야죠. 복잡한 폼 입력을 받는 경우나 필수 입력 조건 같은 여러가지 제약 사항이 있을 땐 여러 validation함수를 체이닝 하거나 validator와 같은 객체를 만들어서 validator 에서 알아서 처리하도록 합니다. 사용하는 입장에선 그 처리과정의 복잡함을 알 필요가 없게 되며, 알아서 처리해주겠지~ 라는 안정감을 얻을 수 있죠.

절대적인 답은 없으며 지금의 복잡도에서 가장 좋은 방식을 딱 필요한 만큼만 취하고 앞서서 추상화 할 필요도 없습니다. 다만 좋은 코드는 변경 하기 좋은 코드이기 때문에 TDD, 테스트 코드와 같은 실천법을 강조하는 겁니다.

현재 수준에서 가장 적절한 방안은 다양한 실험을 통해 결정해보세요. 함수로 빼는게 오히려 의도를 헤치는 경우도 있고 그건 그때 그때 달라서 정답이 없습니다ㅎㅎ

gringrape commented:
> 함수 validation 같은 경우 일단은 각 함수에서 합니다. 다만 공통 validation이 보일 경우 그때 분리하면서 정책을 생각해야죠. 복잡한 폼 입력을 받는 경우나 필수 입력 조건 같은 여러가지 제약 사항이 있을 땐 여러 validation함수를 체이닝 하거나 validator와 같은 객체를 만들어서 validator 에서 알아서 처리하도록 합니다. 사용하는 입장에선 그 처리과정의 복잡함을 알 필요가 없게 되며, 알아서 처리해주겠지~ 라는 안정감을 얻을 수 있죠.
> 
> 절대적인 답은 없으며 지금의 복잡도에서 가장 좋은 방식을 딱 필요한 만큼만 취하고 앞서서 추상화 할 필요도 없습니다. 다만 좋은 코드는 변경 하기 좋은 코드이기 때문에 TDD, 테스트 코드와 같은 실천법을 강조하는 겁니다.
> 
> 현재 수준에서 가장 적절한 방안은 다양한 실험을 통해 결정해보세요. 함수로 빼는게 오히려 의도를 헤치는 경우도 있고 그건 그때 그때 달라서 정답이 없습니다ㅎㅎ

그렇군요. 지금 수준에 가장 적절한 정도의 추상화를 시도해보는 방향으로 하겠습니다.

gringrape committed:
Refactor reducer for simplicity

gringrape committed:
Add test of ListContainer

gringrape committed:
Implement ListContainer

gringrape committed:
Add test of InputContainer

gringrape committed:
Implement InputContainer

gringrape committed:
Refactor InputContainer test

gringrape committed:
Fix App test

couldnt use integration test, because
I needed to configure redux store

gringrape committed:
Add test to InputContainer

for checking if updateTaskTitle event happens correctly

gringrape committed:
Add test to ListContainer

to check if deleteTask event happens correctly

gringrape committed:
Fix InputContainer property pass

gringrape commented:
## 작업내용
1. reducer 리팩토링
2. ListContainer 테스트 추가, 구현
3. InputContainer 테스트 추가, 구현
4. Page 삭제, 기타 정리
### 작업반성
npm test 로 테스트한번 안돌려봐서 통합테스트가 이미 구현되어있는줄도 몰랐습니다.
App 에서 통합테스트 구현하느라 시간소모한 것이 헛된 일이 되었습니다.
기초적인 상태도 확인하지 않는 실수는 되풀이하지 않고 싶네요.

wholemann commented on src/index.jsx:
```diff
@@ -1,9 +1,17 @@
 
+import reducer from './reducer';
+
+const store = createStore(reducer);
```
저희는 store를 store.js 로 분리합니다.

index.jsx에선 store의 정체를 몰라야 유연해집니다ㅎㅎ

gringrape replied:
@wholemann 
유연해진다라는게 변화에 잘 대응한다라는건가요?
여기서는 createStore 라는 구체적인 생성방법이 드러나있어서, 생성방법이 makeStore 라는 식으로
바뀌어 버리면 문제가 되는걸까요?
음... index.jsx 에서는 store 를 어떻게 생성하는지는 관심가질 필요가 없고 가져다 쓰기만 하면되니,
관심사를 분리시킨다라는 측면에서도 생각이 드네요.
유연하다는 의미를 변경의 이유가 같은것이 응집되어 있다? 이렇게 보시는 건가요?

wholemann commented on src/index.jsx:
```diff
@@ -1,9 +1,17 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
+import { Provider } from 'react-redux';
```
보통 2- 3라인은 공백으로 구분해줍니다.

react관련된 것과 redux관련된 것은 구분하는 거죠.

wholemann commented on src/reducer.js:
```diff
@@ -0,0 +1,57 @@
+  tasks: [],
+});
+
+const reducers = {
```
object를 이용해서 if문을 제거하셨군요ㅎㅎ

wholemann commented on src/reducer.test.js:
```diff
@@ -0,0 +1,109 @@
+      });
+    });
+
+    context('with invalid taskTitle', () => {
```
오 이 테스트는 좋군요.

wholemann commented on src/reducer.test.js:
```diff
@@ -0,0 +1,109 @@
+  });
+
+  describe('updateTaskTitle', () => {
+    const previousState = { taskTitle: 'prevTitle' };
```
이런 부분은 

beforeEach 에서 처리 해주는 게 좋습니다. 매 테스트는 독립적으로 작동해야 하니까요.

wholemann commented on src/reducer.test.js:
```diff
@@ -0,0 +1,109 @@
+      const newTaskTitle = 'New Task';
+
+      it('returns new state with new task title', () => {
+        const newState = reducer(previousState, updateTaskTitle(newTaskTitle));
```
```suggestion
        const { taskTitle } = reducer(previousState, updateTaskTitle(newTaskTitle));
```

저라면 이렇게 분해할 거 같아요.

wholemann commented on src/reducer.test.js:
```diff
@@ -0,0 +1,109 @@
+    const previousState = { taskTitle: 'prevTitle' };
+
+    context('with valid taskTitle', () => {
+      const newTaskTitle = 'New Task';
```
new가 붙어있을 필요는 없어요. 어차피 새롭다는 건 우리만 아는 정보이지

이게 전달될 땐 그게 중요하지 않죠. 그냥 전달된 문자열일 뿐이니까요.

new previous와 같은 prefix가 주는 모호함을 몰아내세요ㅎㅎ

굳이 new previous를 붙이지 않아도 '변화' 라는 게 자연히 읽히도록 코딩하는 습관을 만들어야 합니다.

wholemann replied:
```javascript
  describe('updateTaskTitle', () => {
    it('changes new task title', () => {
      const state = reducer({
        taskTitle: '',
      }, updateTaskTitle('New Title'));

      expect(state.taskTitle).toBe('New Title');
    });
  });
```

이러면 충분히 파악이 가능하죠.

wholemann commented on src/reducer.test.js:
```diff
@@ -0,0 +1,109 @@
+    context('with valid taskTitle', () => {
+      const newTaskTitle = 'New Task';
+
+      it('returns new state with new task title', () => {
```
상태를 반환하는 게 핵심이 아니라 'updateTaskTitle'를 통해 changes new task title 하는게 핵심이죠ㅎㅎ

wholemann reviewed (commented):
제가 봤을 때 몇가지 반복되는 아쉬운 점을 남겨드렸습니다. 잘 고민해보시고 혹시 의문이 있다면 언제든 편하게 DM 주세요ㅎㅎ

gringrape committed:
Fix test name of reducer

Focus more on what is tested here
than return value of reducer.
Because it is natural inside reducer test
so I don't need to emphasize it

gringrape committed:
Remove inappropriate prefix on reducer test

gringrape commented:
말씀하신 부분을 반영해보았습니다 :)

- 두괄식 서술로 해당 부분의 핵심적인 테스트 내용이 드러나도록 테스트 이름을 수정했습니다.
- context 가 없어도 케이스가 나눠지는게 보이는 경우는 context 를 삭제하였습니다. 
- 불필요하다고 생각되는 prefix 를 삭제했습니다. 

어떤 값을 여러군데에서 사용하기 위해 변수에 할당할때, 
다른것과 구분하기 위해 new, previous prefix 를 붙이는 것 같습니다.
이 부분은 필요한 경우도 있는것 같아서 조금더 고민해봐야할 것 같습니다.
없어도 변화가 잘 드러나게 하고 꼭 필요한 경우에 쓰는걸 고민해보겠습니다.

gringrape commented:
다 뜯어고치고 싶었으나, 이번주는 좋은 테스트 작성 방향을 잡는 것에 집중하는게 좋을 것 같아서
말씀해주신 부분만 수정해보았습니다. 감사합니다!

gringrape committed:
Extract store out of index.jsx

wholemann commented:
그 전과 지금의 코드를 비교해보시면 어떤 점에서 나아졌는지 비교해보세요ㅎㅎ

자연스러운 말로 읽히는 테스트 코드를 만들기 위해 치열하게 고민해야 합니다!
